contributingGuides/philosophies/PAGINATION.md

```diff
@@ -0,0 +1,113 @@
+# Pagination Philosophy
+## What is pagination?
+Pagination is the art of loading a large amount of ordered data page-by-page, rather than all at once. It is a specialized form of lazy-loading that's suitable for ordered data that displays in a list.
+
+## Why lazy-load data?
+The goal of lazy-loading is to decouple the volume of data a user has access to from the app's performance when they use it. This makes it important to ensure that our largest, high-value customers share the same great app experience as a smaller customer.
+
+It makes sense to lazy-load a resource when the volume of that resource is unbounded, or the upper bound of the resource is too costly to load all at once.
+
+## Pagination vs. lazy-loading
+As stated above, pagination is a specialized subset of lazy-loading for ordered lists of data. The advantage of structured pagination over a more naïve lazy-loading strategy is that it divides data into discrete chunks such that they can be loaded in batches (and that the order of those batches lends itself to a better product experience).
+
+One example of data that's appropriate for pagination is reportActions (chat messages). We can load the 50 most recent messages, then scroll back and send one network request to get the next batch of 50.
+
+One example of data that's appropriate to lazy-load but not necessarily ideal for pagination is avatar images. It's too costly to load all avatar images for all users when the app loads, so we only fetch avatar images that a user wants to see on-demand. Another example might be `personalDetails`.
+
+## How to implement pagination
+1. Clearly define a deterministic sorting order of the data in question.
```

Comment:
NAB bc probably does not need to be pointed out, but are we admitting that we are kind of "without options" philosophy/guidance wise if a list of data is not deterministically sorted? I can only think of one example of course and you know which one it is 😄

Reply:
I should clarify that "deterministic" doesn't mean that the order can't change if the data changes. It just means that if the same data is sorted, the resultant sort order is _exactly_ the same each time.

To explain with an example, the JS array `.sort` function takes a `compareFunction` that's called with two args, `a` and `b`. The return value of `compareFunction` is:

- negative if `a` should come before `b`
- positive if `a` should come after `b`
- zero if `a` and `b` are considered equal

Our sorting order must never return zero - no two items are allowed to be equal.

Reply:
I re-wrote this a bit and I think it's clearer now